### PR TITLE
server: fix error when dedicatingguestvlanrange for physical nw without vlan range

### LINKED ISSUE #6648: Issue when adding VLAN Pool for Guest network
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
Physical network > Guest Network > Adding VLAN Pool
~~~

##### CLOUDSTACK VERSION

~~~
CloudStack 4.16.1.0
~~~

##### CONFIGURATION

Advanced networking configuration with VLANs used on the management, guest and public networks. Using VMWare DVS for guest and public traffic. Public traffic vlan has provisioned in vmware correctly, this issue is limited to guest.

##### OS / ENVIRONMENT

VMware ESXi, 6.7.0, 19898906
VCenter Version 6.7.0 Build 19832280

##### SUMMARY

When adding a pool of VLANs to the guest network under 'Dedicated VLAN/VNI Ranges' we receive the following error: 

Error: (HTTP 530, error code 9999) Index 0 out of bounds for length 0

This is seen when creating through the GUI and also when using Cloud Monkey API Push.

Management server log shows : 

2022-08-17 10:43:17,515 ERROR [c.c.a.ApiServer] (qtp1747352992-77024:ctx-db233219 ctx-fc39cbd8) (logid:bb7bd18a) unhandled exception executing api command: [Ljava.lang.String;@21acc147
java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0

If you refresh the screen, the VLAN pool has been added however when creating an instance the process fails because a VLAN cannot be allocated from the pool. It feels like the VLAN pool is not being fully provisioned on the system. 

##### STEPS TO REPRODUCE

~~~
- Configure a physical network for mgmt using vSwitch0
- Configure a physical network for guest using dvSwitch0 (distributed virtual switch)
- Configure traffic labels
- Create isolated guest network
- Configure VLAN pool
- Create Instance
~~~

##### EXPECTED RESULTS

~~~
VLAN allocated from the pool for isolated guest network
~~~

##### ACTUAL RESULTS

~~~
-Error received when adding VLAN range = 

Error: (HTTP 530, error code 9999) Index 0 out of bounds for length 0

-Error received when creating VM Instance = 

2022-08-17 11:41:26,323 DEBUG [c.c.c.CapacityManagerImpl] (Work-Job-Executor-21:ctx-a2fc1cdc job-1327/job-1328 ctx-b0fa7bd2) (logid:69e5d813) STATS: Failed to alloc resource from host: 4 reservedCpu: 0, requested cpu: 1000, reservedMem: (0 bytes) 0, requested mem: (1.00 GB) 1073741824
2022-08-17 11:41:26,386 DEBUG [o.a.c.e.o.NetworkOrchestrator] (Work-Job-Executor-21:ctx-a2fc1cdc job-1327/job-1328 ctx-b0fa7bd2) (logid:69e5d813) Cleaning up because we're unable to implement the network Ntwk[227|Guest|8]
com.cloud.exception.InsufficientVirtualNetworkCapacityException: Unable to allocate vnet as a part of network Ntwk[227|Guest|8] implement Scope=interface com.cloud.dc.DataCenter; id=5
2022-08-17 11:41:26,777 DEBUG [c.c.c.CapacityManagerImpl] (Work-Job-Executor-21:ctx-a2fc1cdc job-1327/job-1328 ctx-b0fa7bd2) (logid:69e5d813) VM instance {id: "363", name: "i-2-363-VM", uuid: "d3691fe9-7067-4b1c-a40e-ca8d10707d59", type="User"} state transited from [Starting] to [Stopped] with event [OperationFailed]. VM's original host: null, new host: null, host before state transition: Host {"id": "4", "name": "xxxxxxx.xxxxxxx.xxxxxxx", "uuid": "7dfd437b-a8bd-42a5-88df-c76bbc15150d", "type"="Routing"}
2022-08-17 11:41:26,832 ERROR [c.c.v.VmWorkJobHandlerProxy] (Work-Job-Executor-21:ctx-a2fc1cdc job-1327/job-1328 ctx-b0fa7bd2) (logid:69e5d813) Invocation exception, caused by: com.cloud.exception.InsufficientServerCapacityException: Unable to create a deployment for VM instance {id: "363", name: "i-2-363-VM", uuid: "d3691fe9-7067-4b1c-a40e-ca8d10707d59", type="User"}Scope=interface com.cloud.dc.DataCenter; id=5
2022-08-17 11:41:26,832 INFO  [c.c.v.VmWorkJobHandlerProxy] (Work-Job-Executor-21:ctx-a2fc1cdc job-1327/job-1328 ctx-b0fa7bd2) (logid:69e5d813) Rethrow exception com.cloud.exceptio.InsufficientServerCapacityException: Unable to create a deployment for VM instance {id: "363", name: "i-2-363-VM", uuid: "d3691fe9-7067-4b1c-a40e-ca8d10707d59", type="User"}Scope=interface com.cloud.dc.DataCenter; id=5
2022-08-17 11:41:27,057 DEBUG [o.a.c.f.j.i.AsyncJobManagerImpl] (API-Job-Executor-48:ctx-e6681fe2 job-1327) (logid:69e5d813) Complete async job-1327, jobStatus: FAILED, resultCode: 530, result: org.apache.cloudstack.api.response.ExceptionResponse/null/{"uuidList":[],"errorcode":"530","errortext":"Unable to start a VM due to insufficient capacity"}
~~~

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Fixes #6648 
If any of the VLAN from the given range is not found in the database (`cloud.op_dc_vnet_alloc`) then an InvalidParameterValueException will be thrown.
Also, refactors and fixes account check.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

File: server/src/main/java/com/cloud/network/NetworkServiceImpl.java [hunk 1]
```diff
@@ -4068,16 +4068,15 @@ public GuestVlanRange dedicateGuestVlanRange(DedicateGuestVlanRangeCmd cmd) {
...
-            if (allocatedVlans != null && !allocatedVlans.isEmpty()) {
-                for (DataCenterVnetVO allocatedVlan : allocatedVlans) {
-                    if (allocatedVlan.getAccountId() != vlanOwner.getAccountId()) {
-                        throw new InvalidParameterValueException("Guest vlan from this range " + allocatedVlan.getVnet() + " is allocated to a different account."
-                                + " Can only dedicate a range which has no allocated vlans or has vlans allocated to the same account ");
-                    }
-                }
+            List<DataCenterVnetVO> dataCenterVnet = _dcVnetDao.findVnet(physicalNetwork.getDataCenterId(), physicalNetworkId, Integer.toString(i));
+            if (CollectionUtils.isEmpty(dataCenterVnet)) {
+                throw new InvalidParameterValueException(String.format("Guest vlan %d from this range %s is present in the system for physical network ID: %s", i, vlan, physicalNetwork.getUuid()));
```
Comment (weizhouapache) [inline server/src/main/java/com/cloud/network/NetworkServiceImpl.java:4074]: is present or is not present ?

File: server/src/main/java/com/cloud/network/NetworkServiceImpl.java [hunk 2]
```diff
@@ -4068,16 +4068,15 @@ public GuestVlanRange dedicateGuestVlanRange(DedicateGuestVlanRangeCmd cmd) {
...
+            }
+            // Verify guest vlans in the range don't belong to a network of a different account
+            if (dataCenterVnet.get(0).getAccountId() != vlanOwner.getAccountId()) {
```
Comment (weizhouapache) [inline server/src/main/java/com/cloud/network/NetworkServiceImpl.java:4077]: findVnet returns vlans including the allocated and unallocated. 
if vnet is not allocated, the accountId is null, can it be dedicated ?

File: server/src/main/java/com/cloud/network/NetworkServiceImpl.java (hunk 1 above)
Comment (shwstppr) [reply server/src/main/java/com/cloud/network/NetworkServiceImpl.java]: @weizhouapache fixed

File: server/src/main/java/com/cloud/network/NetworkServiceImpl.java (hunk 2 above)
Comment (shwstppr) [reply server/src/main/java/com/cloud/network/NetworkServiceImpl.java]: @weizhouapache added check

File: server/src/main/java/com/cloud/network/NetworkServiceImpl.java [hunk 3]
```diff
@@ -4068,16 +4068,15 @@ public GuestVlanRange dedicateGuestVlanRange(DedicateGuestVlanRangeCmd cmd) {
...
-                    }
-                }
+            List<DataCenterVnetVO> dataCenterVnet = _dcVnetDao.findVnet(physicalNetwork.getDataCenterId(), physicalNetworkId, Integer.toString(i));
+            if (CollectionUtils.isEmpty(dataCenterVnet)) {
+                throw new InvalidParameterValueException(String.format("Guest vlan %d from this range %s is not present in the system for physical network ID: %s", i, vlan, physicalNetwork.getUuid()));
+            }
+            // Verify guest vlans in the range don't belong to a network of a different account
+            if (dataCenterVnet.get(0).getAccountId() != null && dataCenterVnet.get(0).getAccountId() != vlanOwner.getAccountId()) {
+                throw new InvalidParameterValueException("Guest vlan from this range " + dataCenterVnet.get(0).getVnet() + " is allocated to a different account."
+                        + " Can only dedicate a range which has no allocated vlans or has vlans allocated to the same account ");
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/network/NetworkServiceImpl.java:4079]: can you extract to a `verifyGuestVlan()` (or `verifyGuestVlans()` when including the loop in the extraction)?

**Review by DaanHoogland — APPROVED** [review]

File: server/src/main/java/com/cloud/network/NetworkServiceImpl.java (hunk 3 above)
Comment (shwstppr) [reply server/src/main/java/com/cloud/network/NetworkServiceImpl.java]: @DaanHoogland done

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Review by rohityadavcloud — APPROVED** [review]

**Review by weizhouapache — APPROVED** [review]
code lgtm

**Pull request merged**